### ISSUE TITLE
Setup improvements

### ISSUE THREAD

**Issue by nijel** [phpmyadmin/phpmyadmin#12844]
I think we should improve setup to provide more features which are often requested by our users and which would require write access to the filesystem. To make this, the setup needs to be password protected.

- [ ] Request credentials on first run and store them in the configuration file
- [ ] Once configuration exists allow access only based on the credentials
- [ ] Support editing of the configuration file
- [ ] Updater feature
- [ ] Theme download feature

This way it can be both secure and provide features which are often requested. We really can not use MySQL credentials for this as they are not really relevant for the phpMyAdmin configuration access.

**Comment by edwarddekker** [comment]
Please, add also a setup for the database. This would be verry fine.

**Comment by edwarddekker** [comment]
Hello @nijel,

Are you able to announce when will we see the first improvements of the setup?

Regards,
Edward

**Comment by nijel** [comment]
This will definitely not come for 4.7 as we're approaching feature freeze there.

**Comment by Achilles-96** [comment]
Hi @nijel , would this make a good enhancement to add to my proposal?

**Comment by nijel** [comment]
Yes, you can certainly include it.

**Comment by Achilles-96** [comment]
@nijel, do you mean we should store the credentials in raw format.

**Comment by nijel** [comment]
No, they should be hashed, see https://www.owasp.org/index.php/Password_Storage_Cheat_Sheet, not sure which hashing functions are available in PHP, but using Argon2 or PBKDF2 is preferable.

**Comment by Achilles-96** [comment]
Should we allow multiple users or only root user to have this privilege?

**Comment by ibennetch** [comment]
I think even non-root users.

**Comment by Achilles-96** [comment]
So, why can't we just create another table in configuration storage. Also, isn't everyone having access to filesystem risky?
What was there in my mind initially was this. When user wants to setup advanced setup, give him certain instructions such as:
1) Add username and a password to configuration file. (Only users who have access to filesystem would be able to do this and we can check it.)
2) Ask him to give write permission to server to PMA directory.
3) If these conditions are satisfied, we show a form to enter credentials.
4) If the credentials match, we provide him all these additional privileges. This process is independent of mysql users.

**Comment by Achilles-96** [comment]
@nijel @ibennetch Any opinions?

**Comment by ibennetch** [comment]
I don't see the point to creating another table in the configuration storage; do you mean for that to store the credentials for access to this feature? I suppose that would be possible but I'm not fond of polluting the configuration storage table space for only one row of data that could be stored on the file system. On the other hand, I don't particularly know if this information belongs in `config.inc.php`, either. Perhaps @nijel has a specific thought in mind that would clear this aspect up for us. It's good to discuss and plan the merits or weaknesses of any planned implementation, and the only problem I see with using the configuration storage here is that it's an extra table that is largely unnecessary. The other aspect is the security implications — if we put the password in the configuration storage directly, any user with access to that database could discover the password (even if hashed, this is probably a problematic weakness). On the other hand, putting it in the configuration file directly means it's vulnerable to file-level discovery. We have better control over permissions for the file system, though, meaning this is probably the preferred method since we can instruct the user to restrict access to this file.


The rest of your steps sound more or less like what I expect, yes.

**Comment by Achilles-96** [comment]
Thanks for sharing your opinion. @nijel, if you don't think there are any serious problems with asking users to store username and password in a configuration file, I will start implementing this.

**Comment by nijel** [comment]
My original intention was to save password to the configuration file on initial setup and then requiring this password when entering setup. This way it's not bound to MySQL connection at all (what you probably want for configuring phpMyAdmin).

**Comment by Achilles-96** [comment]
So, I will just add one more parameter to configuration, let us say `$cfg['AdvSetupPassphrase']`. Any user who has this passphrase is given access to these features. Does that work?

**Comment by Achilles-96** [comment]
We can also generate the passphrase.

**Comment by Achilles-96** [comment]
@nijel, I think I got what you are trying to say. I started working on it.

**Comment by nijel** [comment]
We already have a javascript code to generate passphrase, that can be reused here as well. But still installing person should be able to choose password.

**Comment by Achilles-96** [comment]
Actually, I still don't understand. Can you please explain the flow which you have in mind.

**Comment by Achilles-96** [comment]
My concerns is putting passphrase as one of the configuration options which can be changed through interface itself is not a good idea. I still feel what I suggested above is reasonable. A minor change though. Re-posting
> 1) Add username and a password to configuration file. (Only users who have access to file-system would be able to do this and we can check it.)
> 2) Ask the user to give write permission to server to PMA directory.
> 3) If these conditions are satisfied, we show a form to enter credentials.
> 4) If the credentials match, we provide the user all these additional privileges. This process is independent of MySQL users.)

Instead of that, we can ask the user to store just the password in a different file, say .pmarc. Only one line is allowed in this file.

We can ask the user to run a script for doing first two steps.

**Comment by nijel** [comment]
Well the whole setup is about changing configuration file, so I don't see problem having setup password as another option there.

For new installs:

- The setup would ask user to enter password during the configuring.
- It would create config.inc.php with the password.

For exiting installs:

- User has to manually add the password to the configuration. This is needed to avoid somebody remotely creating that.

Access to setup:

- If there is no config file, access to setup is allowed for initial setup.
- If there is config without password, access is rejected with link to documentation how to enable it.
- If there is config with password, user has to enter the password prior entering setup.

**Comment by Achilles-96** [comment]
If someone misses initial setup, it is not a big problem till now. But now, anyone can get access to file-system if initial setup is not done.

**Comment by nijel** [comment]
That's true, maybe we should enforce setup if filesystem is writable? That way it's accessible only to first person accessing after installation, what still opens little attack window, but that would be noticed by installing person (assuming that you want to test what you've just installed).

**Comment by Achilles-96** [comment]
What do you think about the method I suggested earlier. Store password in a separate file. This password is never accessible through web. I don't see any problem with that approach. We can ask the user to simply run a script to create password file and to give write permission. Unless this is done, setup is not accessible.

**Comment by nijel** [comment]
I don't think it matters where the password is stored and I personally prefer to have all configuration in single file.

From security view, forcing user to define password first and then allow to use setup is better, but on the other site it's adding another step which might be complicated for some users, so I'm not really sure what is best approach here.

**Comment by nijel** [comment]
I've added it as point for todays meeting, see https://github.com/phpmyadmin/phpmyadmin/wiki/2017-08_Meeting

**Comment by Achilles-96** [comment]
Thanks. I'll try to attend. :)

**Comment by nijel** [comment]
This is what we've agreed on:

- If there is no config.inc.php and it can be written to, redirect user to setup and force him creating config at least with setup password. This will make the attack window minimal as most people will try to access the tool just after installing.
- For existing installs user has to manually add the password to the configuration. This is needed to avoid somebody remotely creating that.

Access to setup:

- If there is no config file, access to setup is allowed for initial setup.
- If there is config without password, access is rejected with link to documentation how to enable it.
- If there is config with password, user has to enter the password prior entering setup.
